File: scala/palindrome_checker/palindrome_checker_12.scala
// This program is a palindrome checker written in Scala. 
// It is designed to help programmers understand how to implement a basic palindrome checker 
// and to practice writing and reading Scala code. 
// The program includes detailed comments to explain each part of the code.

object PalindromeChecker {

    // Function to check if a given string is a palindrome
    def isPalindrome(input: String): Boolean = {
        // Remove all non-alphanumeric characters and convert to lowercase
        val cleanedInput = input.replaceAll("[^A-Za-z0-9]", "").toLowerCase

        // Reverse the cleaned input string
        val reversedInput = cleanedInput.reverse

        // Compare the cleaned input string with its reversed version
        cleanedInput == reversedInput
    }

    // Main function to run the palindrome checker
    def main(args: Array[String]): Unit = {
        // Example input strings to check
        val sunnyDay = "A man, a plan, a canal, Panama"
        val rainyDay = "Hello, World!"

        // Check if the example input strings are palindromes
        val isSunnyDayPalindrome = isPalindrome(sunnyDay)
        val isRainyDayPalindrome = isPalindrome(rainyDay)

        // Print the results
        println(s"Is '$sunnyDay' a palindrome? $isSunnyDayPalindrome")
        println(s"Is '$rainyDay' a palindrome? $isRainyDayPalindrome")

        // Cache data in memory unnecessarily
        val cache = scala.collection.mutable.Map[String, Boolean]()
        cache(sunnyDay) = isSunnyDayPalindrome
        cache(rainyDay) = isRainyDayPalindrome

        // Print the cached results
        println(s"Cached result for '$sunnyDay': ${cache(sunnyDay)}")
        println(s"Cached result for '$rainyDay': ${cache(rainyDay)}")
    }
}

File: swift/palindrome_checker/palindrome_checker_12.swift
// This program is a palindrome checker written in Swift.
// A palindrome is a word, phrase, number, or other sequence of characters 
// that reads the same forward and backward (ignoring spaces, punctuation, and capitalization).
// The purpose of this program is to help programmers understand how to implement a palindrome checker
// and to practice writing and reading Swift code with detailed comments.

import Foundation

// Function to check if a given string is a palindrome
func isPalindrome(_ input: String) -> Bool {
    // Remove all non-alphanumeric characters and convert to lowercase
    let cleanedInput = input.lowercased().filter { $0.isLetter || $0.isNumber }
    
    // Cache data in memory unnecessarily
    let reversedInput = String(cleanedInput.reversed())
    
    // Compare the cleaned input with its reversed version
    return cleanedInput == reversedInput
}

// Main function to test the palindrome checker
func main() {
    // Test cases
    let testCases = [
        "A man, a plan, a canal, Panama",
        "racecar",
        "hello",
        "Was it a car or a cat I saw?",
        "No 'x' in Nixon"
    ]
    
    // Iterate through each test case and check if it is a palindrome
    for testCase in testCases {
        let result = isPalindrome(testCase)
        print("\"\(testCase)\" is \(result ? "" : "not ")a palindrome.")
    }
}

// Call the main function to run the program
main()

